pkg/cmd/server/origin/authenticator.go

```diff
@@ -87,7 +88,8 @@ func newAuthenticator(config configapi.MasterConfig, accessTokenGetter oauthclie
 
 	// OAuth token
 	if config.OAuthConfig != nil {
-		oauthTokenAuthenticator := authnregistry.NewTokenAuthenticator(accessTokenGetter, userGetter, groupMapper)
+		timeouts := authnregistry.NewTokenTimeoutUpdater(accessTokenGetter, oauthClientGetter)
```

Comment:
Timeouts need to be conditional on master config having a default timeout set.  You opt-in to this feature and restart your masters.  We do not support live config changes AFAIK.

Reply:
Nope.
The way I structured it there is no need for any master config.
If you set accessTokenTimeoutSeconds on an OAuthClient then from that point on the feature automatically works.
If you do not do that then this code is ready but does nothing. This means you can add timeouts without having to change configuration/restart masters, which sounds like a much better deal to me. The overhead is so minimal I think it is perfectly acceptable.

Reply:
I disagree.  Someone who uses this probably wants it for all oauth clients, not just the ones they remember to update.  Requiring restart is preferred, that's the way everything else works.

Reply:
The only request I know about wants different timeouts for each client, including the ability to define no timeouts for some clients. Your disagrement is noted, but unless I have evidence that *a* default makes sense I'll just agree to disagree with you on this one.

Reply:
Restarting a master is a big hammer, sometimes it is unavoidable, but in a cluster you would need this plumbing to be always on anyway, otherwise a misconfigured master would cause tokens to timeout on others as it would not participate in keeping them alive. And in general enabling features on the fly is a much better admin experiences. So If I can I am all for *not* having to restart a master to enable a feature when that is not only possible but the easiest way.

Reply:
> a cluster you would need this plumbing to be always on anyway

You do not need it to always be there.  Nor should it always be there.

> misconfigured master would cause tokens to timeout on others as it would not participate in keeping them alive.

You are screwed if you have master configs that are out of sync.

> enabling features on the fly is a much better admin experiences

We purposefully do not do that.  For example, I initially tried to make the token request endpoint's OAuth client update automatically.  But Clayton and David asked me to make it once per server start as we do not support dynamic config changes without server restart.

> So If I can I am all for not having to restart a master to enable a feature when that is not only possible but the easiest way.

Yeah, its the easiest thing to always have something on.  I just believe its fundamentally the wrong thing to do for a feature that less than one percent of people will use.

---

If I was going to use this feature, I would want to know that I can set it up once in master config to turn it on (and never worry about forgetting to set up an OAuth client with a timeout).  Then I can decide to make OAuth clients have different timeouts if I want (this really does not matter much since you can use any token you want with `oc`).